### PR TITLE
Add CuTe DSL HCA backend for DeepSeek V4 sparse MLA

### PR DESCRIPTION
**Pull request by myu-guo**
## Summary

This PR adds a CuTe DSL implementation of DeepSeek V4 FP8 Heavily Compressed Attention (HCA) for Blackwell GPUs.

- Add the SM100/SM103 HCA kernel, two-CTA clusters, split-K reduction, persistent and non-persistent schedulers, and FlashInfer wrapper.
- Expose HCA through `trtllm_batch_decode_sparse_mla_dsv4(..., backend="cute-dsl")` while preserving the existing `backend="auto"` policy.
- Keep the existing API name and add the backend-neutral `batch_decode_sparse_mla_dsv4` alias.
- Support arbitrary SWA token-row order, including ring rotation and wraparound, while keeping the compressed cache paged.
- Add explicit metadata validation, compatibility conversion for combined sparse tables, CUDA Graph-safe reuse of precomputed metadata, correctness tests, and a reproducible benchmark.

The latest kernel revision distributes the four SWA K and V gather4 subtiles across W12-W15 and synchronizes them with the canonical W9/W10 producer states. It also shortens causal-mask, correction, and epilogue live ranges while preserving full-backing causal masking and empty split-K handling.

## Public API and metadata contract

The native CuTe DSL HCA ABI uses two addressing schemes:

- `hca_swa_indices`: INT32 absolute row indices into the flattened SWA cache, shape `[B * Q, 128]`. Each query row may use arbitrary order, rotation, and wraparound.
- `hca_compressed_block_tables`: INT32 physical page IDs into the compressed cache, shape `[B * Q, max_pages]`.
- `hca_seq_lens`: INT32 backing HCA slot counts, shape `[B]`. These drive tile-rounded TMA scheduling and count the fixed 128 SWA slots plus compressed slots; they are neither raw sequence lengths nor effective top-k lengths.
- `swa_topk_lens`: INT32 visible SWA lengths, shape `[B * Q]`, with values in `[0, 128]`.
- `sparse_topk_lens`: INT32 visible total HCA lengths, shape `[B * Q]`.

This replaces the SWA page-table representation used by earlier revisions of this PR. The compressed stream remains page-based.

Every `hca_swa_indices` entry, including masked padding, must contain a legal flattened-cache row because gather4 reads all coordinates before masking. Each compressed block-table row must likewise provide legal page IDs for the complete `hca_seq_lens` footprint rounded up to 128-slot tiles, including slots later masked by a shorter `sparse_topk_lens`.

### Explicit hot-loop path

```python
from flashinfer.mla import trtllm_batch_decode_sparse_mla_dsv4

out = trtllm_batch_decode_sparse_mla_dsv4(
    query=query,
    swa_kv_cache=swa_kv_cache,
    workspace_buffer=workspace,
    sparse_indices=None,
    compressed_kv_cache=compressed_kv_cache,
    sparse_topk_lens=sparse_topk_lens,
    swa_topk_lens=swa_topk_lens,
    backend="cute-dsl",
    hca_swa_indices=hca_swa_indices,
    hca_compressed_block_tables=hca_compressed_block_tables,
    hca_seq_lens=hca_seq_lens,
    hca_use_persistent=True,
)
```

### Compatibility conversion

The native explicit HCA ABI has no page-alignment requirement for SWA rows. Callers whose existing combined `sparse_indices` tensor uses a canonical page expansion for the compressed segment can precompute reusable HCA metadata:

```python
from flashinfer.mla import (
    convert_compressed_page_aligned_sparse_indices_to_hca_metadata,
    trtllm_batch_decode_sparse_mla_dsv4,
)

metadata = convert_compressed_page_aligned_sparse_indices_to_hca_metadata(
    sparse_indices,
    sparse_topk_lens,
    seq_lens,
    swa_kv_cache,
    compressed_kv_cache,
    q_len=query.shape[1],
)

out = trtllm_batch_decode_sparse_mla_dsv4(
    query=query,
    swa_kv_cache=swa_kv_cache,
    workspace_buffer=workspace,
    sparse_indices=None,
    compressed_kv_cache=compressed_kv_cache,
    sparse_topk_lens=sparse_topk_lens,
    swa_topk_lens=metadata.swa_topk_lens,
    backend="cute-dsl",
    hca_swa_indices=metadata.hca_swa_indices,
    hca_compressed_block_tables=metadata.hca_compressed_block_tables,
    hca_seq_lens=metadata.hca_seq_lens,
    hca_is_causal=metadata.hca_is_causal,
    hca_use_persistent=True,
)
```

For `hca_sparse_indices_format="compressed-page-aligned"`, active SWA entries may be arbitrary absolute rows. Only the compressed segment must be a canonical expansion of `page_id * page_size + page_offset`. The old ambiguous `"page-aligned"` tag is rejected.

The inline tagged conversion validates values, allocates metadata, synchronizes the device, and immediately launches the kernel. It is not CUDA Graph capture safe or intended for a latency-sensitive loop. Precompute once and reuse the explicit metadata instead.

## Supported configuration

- SM100 and SM103 implementation target; this revision was validated on SM100 silicon.
- Dense causal query layout `[B, Q, H, 512]`.
- FP8 E4M3 query and KV with BF16 output.
- DeepSeek V4 production configuration of 128 heads; partial final head tiles are masked and tested.
- HND and NHD public cache-layout normalization.
- Optional per-head attention sinks, split-K reduction, and persistent or non-persistent scheduling.
- Non-persistent launches require `B * Q <= 65535`.
- `backend="cute-dsl"` is explicit opt-in; `backend="auto"` remains unchanged.
- CuTe DSL HCA currently requires `hca_is_causal=True`, Python-float BMM scales, and does not support PDL or varlen `cum_seq_lens_q` input.

## Correctness and validation

Validated at final head `d847a2e6a948b283d439d2edcf3cc08226e98594` on an NVIDIA B200:

```text
pytest -q tests/attention/test_cute_dsl_hca_dsv4.py
27 passed
```

The targeted suite covers backend routing and aliasing, HND/NHD forwarding, rotated and wrapped SWA rows, compressed-page-aligned conversion and rejection of the old ambiguous tag, FP8-to-BF16 numerical comparison, sinks, partial head tiles, multiple page sizes, split-K and all-empty contributions, full-backing causal masking, CUDA Graph capture, persistent scheduling, and the W12-W15 four-warp gather contract. Its randomized B=2, Q=2 parity case sends the same query, KV pools, active logical rows, lengths, scales, and sinks to CuTe DSL and TRTLLM-GEN; each backend is checked against the same independent PyTorch reference and against the other backend.

Static checks also passed:

```text
git diff --check
ruff format --check --no-cache
ruff check --no-cache
```

## Final B200 CuTe DSL vs. TRTLLM-GEN comparison

The final comparison was run at `d847a2e6a948b283d439d2edcf3cc08226e98594` on an NVIDIA B200. Here `DKG` denotes the DKG-derived CuTe DSL backend integrated into this checkout. Both backends consume the same logical HCA rows and are selected explicitly; `backend="auto"` remains unchanged.

Each shape was run twice with reversed backend order. The configuration uses 128 heads, head dimension 512, FP8 E4M3 query/cache, BF16 output, shuffled compressed pages, rotated SWA rows, the persistent scheduler, split-K 1, hot-L2 CUDA-event timing, a 512 MiB workspace, 100 warmups, and 200 measured samples per round.

`DKG latency reduction` is `(TRTLLM-GEN - DKG) / TRTLLM-GEN`; positive values favor DKG.

| B | Q | Raw KV | Max active HCA | DKG medians -> mean (us) | TRTLLM-GEN medians -> mean (us) | DKG latency reduction | Diff raw/public/max |
|---:|---:|---:|---:|---:|---:|---:|---:|
| 32 | 4 | 8,192 | 192 | 20.928 / 20.928 -> 20.928 | 22.880 / 22.848 -> 22.864 | 8.47% | 0 / 0 / 0 |
| 32 | 4 | 32,768 | 384 | 22.848 / 22.848 -> 22.848 | 24.896 / 24.864 -> 24.880 | 8.17% | 0 / 0 / 0 |
| 32 | 7 | 8,192 | 192 | 31.296 / 31.200 -> 31.248 | 33.184 / 33.184 -> 33.184 | 5.83% | 0 / 0 / 0 |
| 32 | 7 | 32,768 | 384 | 33.248 / 33.216 -> 33.232 | 39.264 / 39.136 -> 39.200 | 15.22% | 0 / 0 / 0 |
| 1 | 1,024 | 1,048,576 | 8,320 | 616.832 / 618.848 -> 617.840 | 1,144.096 / 1,145.136 -> 1,144.616 | 46.02% | 0 / 0 / 0 |
| 8 | 1,024 | 32,768 | 384 | 833.936 / 833.920 -> 833.928 | 848.160 / 849.088 -> 848.624 | 1.73% | 0 / 0 / 0 |

All six shapes completed both rounds. Raw and public outputs were finite and identical between backends, and logical and encoded active-slot counts matched. Public API execution was checked for correctness but is excluded from this raw-kernel latency table.

### CODE REVIEW

File: flashinfer/cute_dsl/attention/wrappers/batch_hca.py [hunk 1]
```diff
@@ -0,0 +1,621 @@
...
+            compressed_block_tables,
+            compressed_kv_cache,
+            "compressed_block_tables",
+        ),
+    ):
+        invalid_pages = torch.logical_or(tables < 0, tables >= cache.shape[0])
+        if invalid_pages.any().item():
+            raise ValueError(
+                f"{name} values must be physical page IDs in [0, {cache.shape[0]})"
+            )
```
[the remark this answers is not in the document]
Comment (myu-guo) [reply flashinfer/cute_dsl/attention/wrappers/batch_hca.py]: Fixed in `13fd19a`, with one safety adjustment to the suggestion. HCA TMA loads happen before the per-query mask: every row needs the fixed 128-slot window footprint, and compressed pages are determined by the 128-slot tile-rounded `hca_seq_lens` footprint (repeated per query in causal mode), not `sparse_topk_lens`. Validation now ignores only columns beyond that actual load footprint, accepts conventional `-1` tail padding, and rejects missing/invalid active pages. The regression test covers accepted tail padding and both required-window and required-compressed invalid IDs.

File: benchmarks/bench_dsv4_hca_trtllm_vs_dkg.py [hunk 2]
```diff
@@ -0,0 +1,730 @@
...
+    print(
+        f"GPU={torch.cuda.get_device_name()} cc={torch.cuda.get_device_capability()} "
+        f"L2={torch.cuda.get_device_properties(0).L2_cache_size / (1 << 20):.1f} MiB "
+        f"flashinfer={flashinfer_path}"
+    )
+    case_index = 0
+    for raw_seq_len in parse_int_list(args.raw_seq_lens):
+        for batch_size in parse_int_list(args.batch_sizes):
+            if raw_seq_len <= 0 or batch_size <= 0:
+                parser.error("batch sizes and raw sequence lengths must be positive")
```
[the remark this answers is not in the document]
Comment (myu-guo) [reply benchmarks/bench_dsv4_hca_trtllm_vs_dkg.py]: Fixed in `13fd19a`. Both sweep lists are parsed once and rejected with `parser.error` when empty or non-positive, before CUDA availability is checked. Verified `--batch-sizes ,` exits with argparse status 2 and the new diagnostic.

File: flashinfer/cute_dsl/attention/dsa/hca_fp8.py [hunk 3]
```diff
@@ -0,0 +1,3886 @@
...
+            )
+            if tidx == 0:
+                mLSE[blk_coord[0], blk_coord[1], blk_coord[2]] = global_lse
+            # store the scale to shared memory
+            for i in cutlass.range_constexpr(lse_per_thread):
+                split_kv_idx = tidx + i * self.threads_per_warp
+                if cute.elem_less(split_kv_idx, local_split_kv):
+                    smem_lse_scale[split_kv_idx] = cute.math.exp2(
+                        local_lse[i] - global_lse, fastmath=True
+                    )
```
[the remark this answers is not in the document]
Comment (myu-guo) [reply flashinfer/cute_dsl/attention/dsa/hca_fp8.py:1755]: Fixed in `13fd19a`. An all-empty reduction now stores `-inf` LSE and writes zero reduction scales, preserving O=0 without `-inf - -inf` NaNs. The B200 regression captures the private split-K LSE tensor and asserts exact O=0 and all LSE values are `-inf`.

File: flashinfer/cute_dsl/attention/dsa/hca_fp8.py [hunk 4]
```diff
@@ -0,0 +1,3886 @@
...
+            else:
+                for i in cutlass.range_constexpr(page_per_tile_max):
+                    k_idx[i] = common_params.mPT_cmp[
+                        (
+                            cmp_offset * qk_params.tiled_mma_qk.thr_id.shape
+                            + common_params.blk_coord[0]
+                        )
+                        * page_per_tile_max
+                        + i
+                    ]
```
[the remark this answers is not in the document]
Comment (myu-guo) [reply flashinfer/cute_dsl/attention/dsa/hca_fp8.py]: Fixed in `13fd19a`. QK and PV indexing now use each stream own page count for both the table stride and the initialization loop bound; keeping the old max loop bound would still read unused or out-of-range IDs. B200 tests cover both 32/16 and 16/32 window/compressed page-size orderings across two compressed tiles against a gather reference.

File: flashinfer/cute_dsl/attention/dsa/hca_fp8.py [hunk 5]
```diff
@@ -0,0 +1,3886 @@
...
+        if out_dtype not in [cutlass.Float8E4M3FN, cutlass.BFloat16]:
+            return False
+        if acc_dtype != cutlass.Float32 or lse_dtype != cutlass.Float32:
+            return False
+        # Both page sizes must divide mma_qk_tiler[1]; neither may equal 1
+        # (TMA 128B alignment requirement).
+        if mma_qk_tiler_mn[1] % page_size_cmp != 0 or page_size_cmp == 1:
+            return False
+        if mma_qk_tiler_mn[1] % page_size_win != 0 or page_size_win == 1:
+            return False
```
[the remark this answers is not in the document]
Comment (myu-guo) [reply flashinfer/cute_dsl/attention/dsa/hca_fp8.py]: Fixed in `13fd19a`. `can_implement()` now rejects split counts outside `[1, MAX_SPLITS]` and page sizes `<= 1` before modulo arithmetic. CPU contract tests cover split 0, `MAX_SPLITS + 1`, page size 0, page size 1, and a normal supported configuration.

File: flashinfer/cute_dsl/attention/dsa/hca_fp8.py [hunk 6]
```diff
@@ -0,0 +1,3886 @@
...
+        if is_var_split_kv and not is_var_seq:
+            return False
+        if H > 128 or (H < 128 and split_kv != 1):
+            return False
+        if S <= 0:
+            return False
+        if K <= 0:
+            return False
+        return True
```
[the remark this answers is not in the document]
Comment (myu-guo) [reply flashinfer/cute_dsl/attention/dsa/hca_fp8.py:4037]: Fixed in `13fd19a`. The low-level `can_implement()` contract now rejects non-persistent `B * S > 65535`; the public wrapper already had the descriptive launch-grid guard. Tests verify the same oversized shape is rejected for non-persistent and accepted for persistent scheduling.

File: flashinfer/mla/_core.py [hunk 7]
```diff
@@ -1226,15 +1263,117 @@ def trtllm_batch_decode_sparse_mla_dsv4(
...
+        "hca_seq_lens": hca_seq_lens,
+        "hca_use_persistent": hca_use_persistent if hca_use_persistent else None,
+    }
+    unexpected_hca_inputs = [
+        name for name, value in unexpected_hca_inputs.items() if value is not None
+    ]
+    if unexpected_hca_inputs:
+        raise ValueError(
+            f"backend={backend!r} does not accept " + ", ".join(unexpected_hca_inputs)
+        )
```
[the remark this answers is not in the document]
Comment (myu-guo) [reply flashinfer/mla/_core.py:1802]: Fixed in `13fd19a`. Non-HCA backends now include a non-default `hca_is_causal=False` in the unexpected HCA inputs. The test clears the three HCA metadata tensors to isolate this flag and checks that `trtllm-gen` rejects it explicitly.

**Comment by leejnau** [comment]
@myu-guo there is a small merge conflict, could you please address it?

File: flashinfer/cute_dsl/attention/wrappers/batch_hca.py [hunk 8]
```diff
@@ -0,0 +1,660 @@
...
+        window_valid_lens,
+        Float32(1.0),
+        Float32(1.0),
+        sinks,
+        stream,
+        options="--enable-tvm-ffi --opt-level 2",
+    )
+
+
+def _check_tensor(
```
Comment (saltyminty) [inline flashinfer/cute_dsl/attention/wrappers/batch_hca.py:283]: NIT/non-blocking: is there a need to duplicate these helpers compared to using the existing helpers like `utils.py::check_shape_dtype_device`?

**Review by saltyminty — APPROVED** [review]

**Comment by myu-guo** [comment]
Have a bugfix inflight. may need two days.

**Comment by saltyminty** [comment]
CI looks good.
